llm/rust/triton-llm/src/http/service/discovery.rs

```
@@ -13,14 +13,126 @@
+    pub name: String,
+
+    /// Component of the endpoint.
+    pub endpoint: protocols::Endpoint,
+}
+
+pub struct ModelWatchState {
+    pub prefix: String,
+    pub manager: ModelManager,
+    pub drt: DistributedRuntime,
+}
+
+pub async fn model_watcher(state: Arc<ModelWatchState>, events_rx: Receiver<WatchEvent>) {
+    log::debug!("model watcher started");
+
+    let mut events_rx = events_rx;
+
+    while let Some(event) = events_rx.recv().await {
+        match event {
+            WatchEvent::Put(kv) => match handle_put(&kv, state.clone()).await {
+                Ok(model_name) => {
+                    log::info!("added chat model: {}", model_name);
+                }
+                Err(e) => {
+                    log::error!("error adding chat model: {}", e);
+                    // log::warn!(
+                    //     "deleting offending key: {}",
+                    //     kv.key_str().unwrap_or_default()
+                    // );
+                    // if let Err(e) = kv_client.delete(kv.key(), None).await {
+                    //     log::error!("failed to delete offending key: {}", e);
+                    // }
+                }
+            },
+            WatchEvent::Delete(kv) => match handle_delete(&kv, state.clone()).await {
+                Ok(model_name) => {
+                    log::info!("removed chat model: {}", model_name);
+                }
+                Err(e) => {
+                    log::error!("error removing chat model: {}", e);
+                }
+            },
+        }
+    }
+
+    log::debug!("model watcher stopped");
+}
+
+async fn handle_delete(kv: &KeyValue, state: Arc<ModelWatchState>) -> Result<String> {
+    log::debug!("removing model");
+
+    let key = kv.key_str()?;
+    log::debug!("key: {}", key);
+
+    let model_name = key.trim_start_matches(&state.prefix);
+    state.manager.remove_chat_completions_model(model_name)?;
```

Can this fail silently if a model does not exist? Suggestion:

```rust
if let Err(e) = state.manager.remove_chat_completions_model(model_name) {
    log::warn!("Attempted to remove non-existent model: {}. Error: {}", model_name, e);
}
```